src/anndata/_core/merge.py

```diff
@@ -1099,7 +1185,11 @@ def _resolve_axis(
 def axis_indices(adata: AnnData, axis: Literal["obs", 0, "var", 1]) -> pd.Index:
     """Helper function to get adata.{dim}_names."""
     _, axis_name = _resolve_axis(axis)
-    return getattr(adata, f"{axis_name}_names")
+    attr = getattr(adata, axis_name)
+    if isinstance(attr, Dataset2D):
+        return attr.true_index
+    else:
+        return attr.index
```

Comment:
```suggestion
    attr = getattr(adata, axis_name)
    return attr.index
```
(or revert the change, I think), both in conjunction with removing:

```
concat_indices.name = DS_CONCAT_DUMMY_INDEX_NAME
concat_annot.index = concat_indices
```
since what this is trying to handle is when the "true indices" are not used i.e., dummy indices are present.  But we ensure that the "true indices" are used inside of `concat_dataset2d_on_annot_axis`.

Reply:
Could be wrong here, but tests passed locally for me and I think we explicitly test behavior around differing indexes

Reply:
You're right, again still in `tests/lazy` mode mentally, so I was using that as a quick check.